src/librustc_mir/interpret/memory.rs

```diff
@@ -844,6 +845,32 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> Memory<'mir, 'tcx, M> {
         self.get_raw_mut(ptr.alloc_id)?.write_bytes(&tcx, ptr, src)
     }
 
+    /// Writes the given stream of u16s into memory.
+    ///
+    /// Performs appropriate bounds checks.
+    pub fn write_u16s(
+        &mut self,
+        ptr: Scalar<M::PointerTag>,
+        src: impl IntoIterator<Item = u16>,
+    ) -> InterpResult<'tcx> {
+        let src = src.into_iter();
+        assert!(src.size_hint().1.is_some()); // Check that the iterator has an upper bound.
+        let size = Size::from_bytes(u64::try_from(src.size_hint().0).unwrap());
```

Comment:
Please don't call `size_hint` twice.

Reply:
The following snippet [copied from here](https://github.com/rust-lang/rust/blob/cdb50c6f2507319f29104a25765bfb79ad53395c/src/librustc/mir/interpret/allocation.rs#L336) does the right thing:
```rust
        let (lower, upper) = src.size_hint();
        let len = upper.expect("can only write bounded iterators");
        assert_eq!(lower, len, "can only write iterators with a precise length");
```